### PR TITLE
Initial module creation

### PR DESCRIPTION
**Pull request by beardedsamwise**

### CODE REVIEW

File: main.tf [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+
+resource "time_sleep" "wait_5_minutes" {
+  depends_on = [null_resource.previous]
+
+  create_duration = "300s"
+}
+
+// enable API for usage via API keys 
+resource "google_project_service" "api" {
+  project  = "samb-sandbox"
```
Comment (benjaminwestern) [inline main.tf]: Hard coded project ID :)

File: main.tf [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+    server_key_restrictions {
+      allowed_ips = [each.key]
+    }
+  }
+  depends_on = [google_project_service.api]
+}
+
+// create API keys with source hostname based restrctions
+resource "google_apikeys_key" "hostname" {
+  for_each     = toset(var.hostname_restrictions)
```
Comment (benjaminwestern) [inline main.tf:149]: If hostnames is null - what happens?

File: main.tf [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+
+  disable_dependent_services = false
+  disable_on_destroy         = false
+
+  depends_on = [time_sleep.wait_5_minutes]
+}
+
+// create API keys with source ip based restrictions
+resource "google_apikeys_key" "ip" {
+  for_each     = toset(var.ip_restrictions)
```
Comment (benjaminwestern) [inline main.tf:131]: if IP restrictions is null what happens

File: main.tf [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+    create_before_destroy = true
+  }
+}
+
+resource "google_api_gateway_gateway" "api_gw" {
+  provider     = google-beta
+  project      = var.project_id
+  api_config   = google_api_gateway_api_config.api_gw.id
+  gateway_id   = "api-gw-${random_id.default.hex}"
+  display_name = "API GW created by TF module"
```
Comment (benjaminwestern) [inline main.tf:101]: Would be good to pass this in as a variable

File: main.tf [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+  api_id       = "api-gw-${random_id.default.hex}"
+  display_name = "API GW config created by TF module"
+}
+
+resource "google_api_gateway_api_config" "api_gw" {
+  provider      = google-beta
+  project       = var.project_id
+  api           = google_api_gateway_api.api_gw.api_id
+  api_config_id = "config-${local.configmd5}"
+  display_name  = "API GW config created by TF module"
```
Comment (benjaminwestern) [inline main.tf:78]: Would be good to pass this in as a variable

File: main.tf [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+  role    = "roles/apigateway.serviceAgent"
+  member  = "serviceAccount:${google_service_account.default.email}"
+}
+
+// Create API GW + Configs 
+resource "google_api_gateway_api" "api_gw" {
+  provider     = google-beta
+  project      = var.project_id
+  api_id       = "api-gw-${random_id.default.hex}"
+  display_name = "API GW config created by TF module"
```
Comment (benjaminwestern) [inline main.tf:70]: Would be good to pass this in as a variable

File: main.tf (hunk 3 above)
Comment (beardedsamwise) [reply main.tf:131]: No API keys are created with IP restrictions.

File: main.tf (hunk 2 above)
Comment (beardedsamwise) [reply main.tf:149]: No API keys are created with hostname restrictions.

File: main.tf (hunk 4 above)
Comment (beardedsamwise) [reply main.tf:101]: Ez, will do this.

File: main.tf (hunk 5 above)
Comment (beardedsamwise) [reply main.tf:78]: Ez, will do this.

File: main.tf (hunk 6 above)
Comment (beardedsamwise) [reply main.tf:70]: Ez, will do this.

File: main.tf (hunk 2 above)
Comment (beardedsamwise) [reply main.tf:149]: I've noted in the docs that the user MUST specify either an IP or hostname to get a key.

File: main.tf (hunk 3 above)
Comment (beardedsamwise) [reply main.tf:131]: I've noted in the docs that the user MUST specify either an IP or hostname to get a key.

File: main.tf (hunk 1 above)
Comment (beardedsamwise) [reply main.tf]: Zoinks! Will fix.

**Review by benjaminwestern — APPROVED** [review]
LGTM!

**Pull request merged**